Completed orders should break items out by Shop

zenweasel opened the pull request:
Closes #2308 and #2245

My first React component replacement from scratch so I open to input on layout, structure, etc. Layout is responsive thanks to help from @kieckhafer. Also first time I have done any HTML of significance so there might be problems there.

### To Test

#### Single Shop (for just the completed page functionality)

1. Place an order
1. Proceed to the completed order page

#### Multi-shop

**[Update]** Slightly shorter instructions now that you can see products from both shops w/o switching shops.

1. Login as admin (Marketplace admin)
1. Set up shipping (I've only been testing with Flat Rate but YMMV)
1. Enable/Configure Stripe (keys are in `reaction.json`)
1. Click on "Shop" settings in the Dashboard
1. Click on "Marketplace"
1. Enter a name and email in the "Invite Owner Form" and click "Send Invitation"
1. Copy the URL from the invite email
1. Open that URL in an incognito window (you will want to keep your logged in session)
1. Open "My Shop Settings"
1. Click on "Start Accepting Payment". You will be taken to Stripe's site
1. **Don't fill out this form**. Click on the "Skip this account form" link at the top, this will return you to your local site and Stripe is now connected 
1. While still a Shop Admin, create a product and publish
1. In your non-incognito browser window as Marketplace admin, place a Reaction item in your cart
1. Add your newly created Shop product to your cart
1. Checkout as normal

**Note that "Estimated Ship Date" is not implemented on the back end so that's why it doesn't display here.

**Another Note** There is currently an issue with the way that discounts are being stored on multi-shop orders so that discounts are doubled (once per shop).

zenweasel commented:
Make up your mind ESLint:
```
  42:1   error    Expected indentation of 10 spaces but found 18                                 indent
  42:19  warning  The closing bracket must be aligned with the opening tag (expected column 18)  react/jsx-closing-bracket-location
```

mikemurray commented on imports/plugins/core/checkout/client/components/addEmail.js:
```diff
@@ -0,0 +1,92 @@
+import { Components } from "@reactioncommerce/reaction-components";
+
+
+class AddEmail extends React.Component {
```
`import React, { Component } fron "react"` and use `Component` in place of `React.Component`

zenweasel replied:
So we have both `Component` and `Components` in the namespace. That seems unfortunate?

mikemurray replied:
That is unfortunate. Maybe change it, maybe don't. Thinking about consistency, especially in the new jsdocs.

zenweasel replied:
Yeah, you're right

zenweasel commented:
@mikemurray I get that error on `marketplace` from time to time on all branches, I don't think it's related to this PR

mikemurray commented:
@zenweasel I figured. Other than those errors, your stuff appears to be good.

aaronjudd commented on imports/plugins/core/checkout/client/components/completedOrderPaymentMethods.js:
```diff
@@ -0,0 +1,35 @@
+const creditCardClasses = {
+  VISA: "fa fa-cc-visa",
+  MAST: "fa fa-cc-mastercard",
+  DISC: "fa a-cc-discover",
```
typo?

aaronjudd commented on imports/plugins/core/checkout/client/components/completedOrderPaymentMethods.js:
```diff
@@ -0,0 +1,35 @@
+import { registerComponent } from "@reactioncommerce/reaction-components";
+
+const creditCardClasses = {
+  VISA: "fa fa-cc-visa",
```
Somehow I wish there were aliases to these in .less or someplace...

kieckhafer replied:
Should these be showing up?

zenweasel replied:
They are showing up for me and for Ryan at least (from his screenshots)

aaronjudd commented on imports/plugins/core/checkout/client/components/completedOrderItem.js:
```diff
@@ -0,0 +1,30 @@
+  if (handleDisplayMedia(item)) {
+    image = handleDisplayMedia(item).url();
+  } else {
+    image = "/resources/placeholder.gif";
```
wish the wasn't hardcoded

zenweasel replied:
This is what we do pretty much everywhere isn't it? Do we have another solution? Maybe a `Reaction.getPlaceholder` method?

spencern replied:
Probably worth permitting shop owners to upload their own placeholder at some point and providing a fallback to a default placeholder if a custom placeholder is not available

aaronjudd replied:
@rymorgan "default placeholder" could / should be one of the UI settings

rymorgan replied:
@aaronjudd Default place holder for Visa, Mastercard etc?

zenweasel replied:
@rymorgan I think he means for the default **Product** image. Rather than using the one included with RC.

aaronjudd commented on private/data/i18n/en.json:
```diff
@@ -429,7 +429,15 @@
         "trackYourDelivery": "Order updates will be sent to",
         "noOrdersFound": "No orders found.",
-        "adminMessage": "You are viewing orders as an administrator"
+        "adminMessage": "You are viewing orders as an administrator",
```
Let's go ahead and add the LingoHub translations in this PR as well.. don't you think?

aaronjudd commented on server/publications/collections/cart.js:
```diff
@@ -67,6 +68,32 @@ Meteor.publish("Cart", function (sessionId, userId) {
+    return item.product._id;
+  }));
+
+  const variantIds = _.uniq(cartItems.map(function (item) {
```
Isn't there an ES6 version of ._unique?

rymorgan commented:
Design review notes: 

![cart_completed-notes](https://user-images.githubusercontent.com/20409254/29330504-38655a76-81ae-11e7-8653-793d4e7e10f1.png)

kieckhafer commented:
Aside from what @rymorgan said, it seems a bit wide on smaller screens... do we want to add some padding on the sides when it's all one column?
![cart_completed_and_collabscreen_js_ ___projects_reaction_foray-reaction](https://user-images.githubusercontent.com/4482263/29336470-b625577c-81c2-11e7-9dd0-12ca54e7d6ec.png)

zenweasel commented:
@rymorgan I don't see the "Your Cart" on the design spec should it just be the same as the other section titles? Should this say "Your Cart" or "Your Order" since this is the completed order page?

spencern commented:
Was able to see this on a checkout through auth.net
a few notes:
I had to refresh the page after getting to checkout-completed to get the spinner to go away. I remember you mentioning that you'd seen that in marketplace as well though, so probably not introduced here.

@rymorgan I was just looking through the designs and noticed that there's only a spot for one invoice there, for multi-shop orders is the idea that we just combine all invoices? I worry that might confuse the customer especially when they see multiple charges on their account. I'll make notes in zeplin as well.

Ryan and Erik pointed out the alignment issues already, so I won't dig there.

rymorgan commented:
@zenweasel - Yea, I think we should add that for consistency w/t the cart. You're right it wasn't in the design spec. I've added it for clarity. 

https://zpl.io/2pnvEya

zenweasel commented:
I believe I have fixed:

1. Alignment issues
1. Added "Your Cart" header
1. Increased margin for narrow widths
1. CC icon not always showing up

1.

zenweasel commented:
@spencern @rymorgan I am just noticing now (re-reading the original ticket to make sure I wasn't missing anything) that the spec says we want to have subtotals for each shop but that wasn't in the design spec. Is that something we need to do now?

spencern commented:
@zenweasel @rymorgan I mentioned this to Ryan and added notes to the zeplin file. 

We have invoices for each payment that comes through. We have payments for each shop represented (there may be future exceptions to this, not sure) .

I think we should discuss whether we should identify multiple invoices for each shop rather than just subtotal so that it matches up with a customer's credit card statement.

rymorgan commented:
@zenweasel @spencern -- Taking a look at this is in my queue. Not sure if we should push this PR out and create another issue for it while I design it? Up to y'all.

zenweasel commented:
@rymorgan I'm all for merging this in, we have several tickets still outstanding for this page where can still make changes.

zenweasel commented:
@spencern @rymorgan I feel like just having subtotals here and maybe making it even more explicit ("$XX.XX will be charged from ThisShop") would be more clear than having multiple invoices (unless you are thinking something different from what I am thinking for invoices). For me as a consumer I would start to feel like you are putting the complexity on me.

spencern commented:
That may be the case, I can understand that argument for sure.
Will it be confusing to lump all taxes together even if they are charged a multiple rates?

zenweasel commented:
My thinking is each shop would have a subtotal that included tax (so it would reflect the total amount charged). Is what you are thinking that we would not have the "Grand Total" there and only Shop-subtotals?

spencern commented:
Subtotals don't generally include tax as far as I'm aware. I'm not really sure what the rules are around this, but I think lumping tax into a subtotal might be frowned upon.

I think frequently the tax rate is listed along side the taxes charged. That's the main reason I'm concerned about not splitting.

zenweasel commented:
I think the subtotal could include a "merchandise total" (or whatever) and then a tax amount for each shop, and then the actual "charged total" (separate from the Grand Total)

rymorgan commented:
Here's a mock with messaging for multiple brand/charge payment. https://zpl.io/VKXwn7a

zenweasel commented:
@rymorgan @spencern This has been updated with the by Shop subtotals according to Ryan's comp

rymorgan commented:
Looking really good, @zenweasel is just slaying some HTML. 
Two small tweaks and then I think the UI is GTG. 
<img width="1466" alt="fullscreen_8_23_17__2_56_pm" src="https://user-images.githubusercontent.com/20409254/29640469-f88b1c26-8813-11e7-83eb-7f1edbe8bfe2.png">

zenweasel commented:
@aaronjudd @spencern Ready to merge?

aaronjudd commented:
Attempting to checkout using Shippo+PayPalExpress, I get:

```
TypeError: Cannot read property 'substring' of undefined
    at CompletedOrderPaymentMethod (app.js:38855)
    at modules.js:108224
    at measureLifeCyclePerf (modules.js:107994)
    at ReactCompositeComponentWrapper._constructComponentWithoutOwner (modules.js:108223)
    at ReactCompositeComponentWrapper._constructComponent (modules.js:108198)
    at ReactCompositeComponentWrapper.mountComponent (modules.js:108106)
    at Object.mountComponent (modules.js:100937)
    at ReactDOMComponent.mountChildren (modules.js:107292)
    at ReactDOMComponent._createInitialChildren (modules.js:104491)
    at ReactDOMComponent.mountComponent (modules.js:104310)
```

aaronjudd commented:
Get the following error on `cart/completed` with Example Payment + Shippo:
```
modules.js?hash=ec5a684de011ed4139d36489b5beb81c37c2df0e:85667 Warning: Unknown prop `i18bnKey` on <span> tag. Remove this prop from the element. For details, see https://fb.me/react-unknown-prop
    in span (created by Reaction(Translation))
    in Reaction(Translation) (created by AddEmail)
    in form (created by AddEmail)
    in AddEmail (created by Reaction(CompletedOrder))
    in div (created by Reaction(CompletedOrder))
    in div (created by Reaction(CompletedOrder))
    in Reaction(CompletedOrder) (created by Tracker(CompletedOrder))
    in Tracker(CompletedOrder) (created by withProps(Tracker(CompletedOrder)))
    in withProps(Tracker(CompletedOrder))
```

aaronjudd commented:
@rymorgan some design questions:
1) See question of case.  We should discuss case, not sure if we've documented what we'd like to see or not. Previous standard was `Cart total`vs `Cart Total`.
![basic_reaction_product_and_completed_orders_should_break_items_out_by_shop_by_zenweasel_ _pull_request__2645_ _reactioncommerce_reaction](https://user-images.githubusercontent.com/439959/29742877-ff3ccc0c-8a3b-11e7-9fbb-c8cdae4a03b3.png)

![cart_completed](https://user-images.githubusercontent.com/439959/29742807-8efebef6-8a3a-11e7-903f-b7446af33281.png)
I'm not sure what `placed by` means?  shouldn't this be `Fulfilled by`? or `Placed with`

![cart_completed](https://user-images.githubusercontent.com/439959/29742949-09fe0e66-8a3d-11e7-9675-b54f5ebaba97.png)
The previous version of this page, well, feels a bit more refined I think, not sure if it's the boxes, all the bold font, _perhaps the all gray background_, or the lack of the green CTA.

aaronjudd commented on server/publications/collections/cart.js:
```diff
@@ -67,6 +67,33 @@ Meteor.publish("Cart", function (sessionId, userId) {
   return Cart.find(cartId);
 });
 
+
```
@machikoyasuda It'd be useful to have all the publications documented using jsdoc.  I think I just never started them because I wasn't sure the best format. Maybe you could/doc/suggest the best format?

aaronjudd commented on imports/plugins/core/checkout/client/components/addEmail.js:
```diff
@@ -0,0 +1,92 @@
+import { Orders } from "/lib/collections";
+import { Reaction, i18next } from "/client/api";
+import { Components } from "@reactioncommerce/reaction-components";
+
```
We should be adding docs to all new components: https://github.com/reactioncommerce/reaction-jsdoc#documenting-react-components

aaronjudd commented:
@rymorgan The order/completed summaries... don't show discount codes anywhere.. anymore... (and it should, I think).

aaronjudd commented:
Previous cart logic was, that if a value in the line items is 0, then it should not display.

<img width="455" alt="cart_completed" src="https://user-images.githubusercontent.com/439959/29743812-b9814776-8a4d-11e7-9abb-a0fecfa37039.png">

rymorgan commented:
@aaronjudd, good catch on title casing vs regular. I prefer title casing to regular casing but at the same time am open to argument for regular case. I've done a quick informal look around and it seems like there isn't much of a standard for what people do. I brought it up with @machikoyasuda and @sophiehe to see where they stand. They both are on team title case. Are you good with that? 

Also, I agree about the awkward wording around Order placed by, I've made that adjustment in the design spec as well. Updated 'Fulfilled by'

I'm looking into the discount and the CTA field. I didn't intentionally remove discounts, thought we'd keep handling it the same way as before.

rymorgan commented:
@aaronjudd I'm seeing discounts. Am I missing something? 

![cart_checkout-discounts](https://user-images.githubusercontent.com/20409254/29796664-4542d5da-8c07-11e7-96ee-10db8159662d.png)

rymorgan commented:
@zenweasel A couple of design tweaks: 

Center contents of the page: 
![cart_completed](https://user-images.githubusercontent.com/20409254/29796761-d5e1e126-8c07-11e7-9c26-81fd1a3cba49.png)

Also per the feedback from @aaronjudd I made the boxes white so that they would stand out from the grey. I also removed some of the bolding, to clean up the hierarchy. 

https://zpl.io/VKXwn7a

spencern commented:
@zenweasel @aaronjudd is this getting close?

zenweasel commented:
@spencern I hope to have this wrapped up today

zenweasel commented:
@aaronjudd @spencern 

### Items that have been fixed:

* Various styling changes based on Updated comps
* Styling of Email input when user is anonymous
* Displaying Discounts
* Not displaying Tax, etc. when 0
* Added JSDocs for components

rymorgan commented:
@zenweasel From a design standpoint, this looks good to me. One small thing and then I'm good on this one. 

![cart_completed](https://user-images.githubusercontent.com/20409254/29940411-fe1c86c6-8e43-11e7-9032-16f4aadec098.png)

spencern commented:
Checking back in on this.
@aaronjudd - are your requested changes satisfied now?
@rymorgan are we good here now?

Tests are failing and branch is out of date.

rymorgan commented:
@spencern As long as that last tiny css change is in there. I'm good yes. (see my previous comment)

spencern commented:
One small thing and one big thing.
**Small:**
I think we should probably keep the shipping name next to the shipping address
![image](https://user-images.githubusercontent.com/1203639/30067831-2a078c8a-9219-11e7-8dd5-1ee8851f3c3b.png)

**Big:**
After placing one order as a new user, I am unable to add products to my cart. There are no errors in the client or server consoles
http://recordit.co/epPjELF45R

Doing a hard refresh shows the products in the cart
![image](https://user-images.githubusercontent.com/1203639/30068024-b82b86ec-9219-11e7-802b-743d9349d8d1.png)

spencern commented:
Overall this feels pretty good though, and is much more marketplace ready than what we had before.

![image](https://user-images.githubusercontent.com/1203639/30068193-3c89f0a4-921a-11e7-840b-b40ea6cd6ff4.png)

rymorgan commented:
@zenweasel @spencern Totally agree that the Name should be in the shipping box. Good catch. I added to my zeplin mocks.

spencern commented:
@rymorgan 
While we're discussing copy/design on this. 

"Quantity Total" feels like a really weird way to say the total number of items in an order. I realize that it's itemized in the box to the left, which helps, but it still feels weird to me.

rymorgan commented:
@spencern I don't disagree, I took that from what was there because @aaronjudd wanted me to use the existing language. I think 'Quantity' would work though.

zenweasel commented:
@spencern What is the issue you are seeing as I don't see it? Can you provide some steps to reproduce?

zenweasel commented:
@spencern I'm not able to replicate the "Cart Empty" issue with either a single shop or multi-shop checkout. Is there something specific I should be doing?

![checkout](https://user-images.githubusercontent.com/72819/30089027-ad572244-92dc-11e7-965c-6a9ce43fd043.gif)

I'll keep trying though

zenweasel commented:
Ok, I was finally able to reproduce the bug and it's something that we fixed a long time ago with this code:

```
  const sessionId = Session.get("sessionId");
  const userId = Meteor.userId();
  const cartSub = Reaction.Subscriptions.Cart = Meteor.subscribe("Cart", sessionId, userId);
  cartSub.stop();
  Reaction.Subscriptions.Cart = Meteor.subscribe("Cart", sessionId, userId);
```
It seems to be caused by the fact that we delete that Cart record and for some reason the subscription doesn't update (possibly the OpLog is tied to that original ID). So stopping and resubscribing used to fix it.

However, putting this same code in the container does not seem to resolve this issue as the new subscription is not going into place. Continuing to try and see how I can fix this. Super painful because it only happens the first time after a full reset.

zenweasel commented:
@spencern Ok, so that bug should be fixed now

rymorgan commented:
We should make it consistent here then. (this is where it from)

<img width="440" alt="cart-labels" src="https://user-images.githubusercontent.com/20409254/30122317-da94e8f8-92e2-11e7-9562-bef2520a58ae.png">

aaronjudd commented:
@rymorgan and where you got it from is inconsistent..

spencern commented:
@aaron is this language a blocker for you here or should I merge and create a new issue for updating copy?

aaronjudd commented:
@spencern let's create an issue for label consistency, and we can tackle as one last design review before we call release the first marketplace. I'll give this full reviewing, ignoring labels in 10..

rymorgan commented:
@spencern @aaronjudd -- I'll go through all my zeplin docs and update the label to "Items in cart" and also update everything to sentence case.
